README
Copyright (C) 2016 by Marian HackMan Marinov

Linux-Setns
========================

setns is used to change the current namespace of the process by specifying a file descriptor to a currently available namespace.

So you must provide a file descriptor and a type (CLONE_ALL, CLONE_NEWIPC, CLONE_NEWNET, CLONE_NEWPID, CLONE_NEWUSER, CLONE_NEWUTS and CLONE_NEWCGROUP).

On successfull execution your process will now be in the desired namespace.
In case of failure, error message is printed to STDOUT.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module does not require any 3rd party libraries, but it does require your machine to have SYS_setns in <sys/syscall.h> for build.
And also it requires fairly recent kernel to work.

The interface to setns() changed between 0.05 and 1.00.

Version before 1.00 does not work on machines with older glibc.

CLONE_NEWCGROUP is available only on 4.6 and newer kernels.

BUGS AND SUGGESTIONS
	contact me at hackman at cpan.org

REPOSITORY
	http://github.com/hackman/linux-setns

COPYRIGHT AND LICENCE

This code is under GPLv2.
A proper copy of the license can be found in the LICENSE file located in the current directory.

Copyright (C) 2014 by Marian HackMan Marinov

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.1 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Linux::Setns.

2.0   Tue Nov  1 04:58:35 2016
    - Added CLONE_NEWCGROUP flag

1.04  Thu Oct  6 02:23:41 2016
    - Removed wrongly added files in the archive

1.03  Wed Oct  5 23:08:40 2016
    - Introduced META.json
    - Added LICENSE and META.json to the MANIFEST
	- Added required perl version to both META.json and META.yml

1.02  Thu Aug 18 15:45:00 2016
	- Added CLONE_NEWNS, CLONE_NEWPID and CLONE_NEWUSER
	- Fixed the documentation in Setns.pm

1.01  Thu Aug 18 15:05:00 2016
	- Changed the license to Perl(thanks to Mohammad S Anwar - https://github.com/Manwar)
	- Update the documentation of the Perl module to reflect the changes done in 1.01

1.00  Wed Aug 17 19:20:10 2016
	- Almost complete rewrite of the function!
	- Changed the interface, to setns(). Now instead of file descriptor we directly give it path to a namespace file. This removes the need to convert Perl FD into C FD.
	- Improved the error handling and now we catch all setns() errors
	- Replaced __NR_setns with SYS_setns. In order for the change to work I also changed the included headers
	- Simplified the setns() function in the XS code
	- Replaced the PPCODE with CODE and now I properly return the RETVAL from the XS code
	- Added new test for invalid path

0.05  Fri Jun 03 10:52:49 2016
	- Replaced the Perl 5.18 requirement with Perl 5.10

0.04  Mon Sep 01 00:57:51 2014
	- Added repository and lincense links

0.03  Sun Aug 17 02:40:05 2014
    - Updated the example, as it had errors(the use statment was for unshare, not setns)

0.02  Mon Jul 28 04:08:09 2014
	- Updated the version and removed some commented code
	- Added basic META.yml file

0.01  Thu Jul 24 08:08:22 2014
	- original version; created by h2xs 1.23 with options
		-A -n Linux::Setns


build.sh
#!/bin/bash
cd ~/Projects/perl-containers/

proj='Linux-Setns'
ver=$(awk '/^version:/{print $2}' $proj/META.yml)
excludes='t.pl Setns.o Setns.so Makefile.old .git build.sh up-version.sh convert-yaml-to-json.pl Makefile'

if [ -z "$ver" ]; then
	echo "Unable to get $poj version"
	exit 1
else
	echo "Building project $proj version $ver"
fi


if [ ! -d $proj ]; then
	echo "Unable to find $proj dir"
	exit 2
fi

for i in $excludes; do
	exclude_option="$exclude_option --exclude $i"
done

# Update the copyright year
year=$(date +%Y)
sed -i "/Copyright.*Marian/s/[0-9]\+/$year/" $proj/LICENSE $proj/lib/Linux/Setns.pm
# Update the META.json
$proj/convert-yaml-to-json.pl $proj/META.yml

if [ -f $proj-${ver}.tar.gz ]; then
	rm -f $proj-${ver}.tar.gz
fi

mv $proj ${proj}-$ver
tar $exclude_option -czf $proj-${ver}.tar.gz ${proj}-$ver
mv ${proj}-$ver $proj



Makefile.PL
use 5.010001;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Linux::Setns',
    VERSION_FROM      => 'lib/Linux/Setns.pm', # finds $VERSION
    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Linux/Setns.pm', # retrieve abstract from module
       AUTHOR         => 'Marian HackMan Marinov <hackman@cpan.org>') : ()),
    LIBS              => [''], # e.g., '-lm'
    DEFINE            => '', # e.g., '-DHAVE_SOMETHING'
    INC               => '-I.', # e.g., '-I. -I/usr/include/other'
	# Un-comment this if you add C files to link with later:
    # OBJECT            => '$(O_FILES)', # link all the C files too
);


lib/Linux/Setns.pm
package Linux::Setns;

use 5.010001;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Linux::Setns ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	setns CLONE_ALL CLONE_NEWNS CLONE_NEWIPC CLONE_NEWNET CLONE_NEWUTS CLONE_NEWPID CLONE_NEWUSER CLONE_NEWCGROUP
);

our $VERSION = '2.00';

use constant {
	CLONE_ALL => 0,
	CLONE_NEWNS => 0x00020000,
	CLONE_NEWIPC => 0x08000000,
	CLONE_NEWNET => 0x40000000,
	CLONE_NEWUTS => 0x04000000,
	CLONE_NEWPID => 0x20000000,
	CLONE_NEWUSER => 0x10000000,
	CLONE_NEWCGROUP => 0x02000000
};

require XSLoader;
XSLoader::load('Linux::Setns', $VERSION);

# Preloaded methods go here.

sub setns {
	my $ret = setns_wrapper($_[0], $_[1]);
	if ($ret == 0) {
		return 1;
	} elsif ($ret == 1) {
		print STDERR "Error: setns() The calling thread did not have the required privilege (CAP_SYS_ADMIN) for this operation\n";
	} elsif ($ret == 2) {
		print STDERR "Error: setns() Unable to open file $_[0]\n";
	} elsif ($ret == 9) {
		print STDERR "Error: setns() FD is not a valid file descriptor\n";
	} elsif ($ret == 12) {
		print STDERR "Error: setns() Cannot allocate sufficient memory to change the specified namespace\n";
	} elsif ($ret == 22) {
		print STDERR "Error: setns() FD refers to a namespace whose type does not match that specified in nstype, or there is problem with reassociating the the thread with the specified namespace\n";
	}
	return 0;
}

1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

Linux::Setns - Perl extension for switching the current process namespace to another namespace pointed by a path to the ns file descriptor.

=head1 SYNOPSIS

	use Linux::setns qw(setns CLONE_ALL CLONE_NEWIPC CLONE_NEWNET CLONE_NEWUTS CLONE_NEWUSER CLONE_NEWPID);

	die "setns() requires root privileges\n" if $>;

	setns("/proc/PID/ns/mnt", CLONE_ALL);
	# now your process is in the same namespaces(IPC,NET,UTS,PID,USER,MOUNT) as the /proc/PID/ns/mnt 

	# If you want to change only one of your namespaces you can use any of the bellow examples:

	# Switch your current Mount namespace to the one pointed by /proc/PID/ns/mnt
	setns("/proc/PID/ns/mnt", CLONE_NEWNS);

	# Switch your current IPC namespace to the one pointed by /proc/PID/ns/ipc
	setns("/proc/PID/ns/ipc", CLONE_NEWIPC);

	# Switch your current Network namespace to the one pointed by /proc/PID/ns/net
	setns("/proc/PID/ns/net", CLONE_NEWNET);

	# Switch your current UTS namespace to the one pointed by /proc/PID/ns/uts
	setns("/proc/PID/ns/uts", CLONE_NEWUTS);

	# Switch your current Pid namespace to the one pointed by /proc/PID/ns/pid
	setns("/proc/PID/ns/pid", CLONE_NEWPID);

	# Switch your current User namespace to the one pointed by /proc/PID/ns/user
	setns("/proc/PID/ns/user", CLONE_NEWUSER);

	# Switch your current Cgroup namespace to the one pointed by /proc/PID/ns/user
	setns("/proc/PID/ns/user", CLONE_NEWCGROUP);

=head1 DESCRIPTION

This trivial module provides interface to the Linux setns system call. It
also provides the CLONE_* constants that are used to specify which kind of
namespace you are entering. Also a new CLONE_ALL constat is provided so you
can join/switch to any type of namespace.

The setns system call allows a process to 'join/switch' one of its namespaces
to namespaces pointed by a file descriptor(usually located in /proc/PID/ns/{ipc,mnt,net,pid,user,uts}).

Note: keep in mind that using any specific CLONE_NEW* constant will fail if the FD path 
you gave is not of that type.

RETRUN VALUE
	1 on success
	0 on failure


=head2 EXPORT

 setns			- the subroutine

 CLONE_ALL		- flag that tells that the path can be of any namespace type
 CLONE_NEWNS	- when this flag is used the path must be from another Mount namespace
 CLONE_NEWIPC	- when this flag is used the path must be from another IPC namespace
 CLONE_NEWNET	- when this flag is used the path must be from another Network namespace
 CLONE_NEWUTS	- when this flag is used the path must be from another UTS namespace
 CLONE_NEWPID	- when this flag is used the path must be from another PID namespace
 CLONE_NEWUSER	- when this flag is used the path must be from another User namespace
 CLONE_NEWCGROUP - when this flag is used the path must be from another Cgroup namespace


=head1 SEE ALSO

setns(s) Linux man page.

=head1 AUTHOR

Marian HackMan Marinov, E<lt>hackman@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2016 by Marian HackMan Marinov

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.1 or,
at your option, any later version of Perl 5 you may have available.


=cut
